### DISCUSSION TITLE
When using zod resolver, z.object() does not work with anything other than z.string and refine().

### DISCUSSION THREAD

**Discussion by apn-atobe** [react-hook-form/react-hook-form#8809] [Q&A] [answered]
I have created a [sample](https://codesandbox.io/s/jovial-colden-ew7l55?file=/src/App.js).

If you insert the following z.object() into the resolver, you will see that the console.log in refine() stops working.
We know that z.boolean(), etc. works the same way as z.number().

Is there some kind of rule for using resolver?

```
const reqBodyScheme = z
  .object({
    password: z.string().min(1, { message: "required" }),
    passwordConfirm: z.string().min(1, { message: "required" })),
    NumberInput: z.number() // <- If you comment this out and use only z.string(), the refine works
  })
  .refine(
    (data) => {
      // Not working
      console.log(data);
      return data.password === data.passwordConfirm;
    },
    {
      message: "Passwords don't match",
      path: ["passwordConfirm"], }
    }
  );
```

Maybe I'm using it wrong...?

codeSandbox: https://codesandbox.io/s/jovial-colden-ew7l55?file=/src/App.js

**Comment by Moshyfawn** [comment]
This behavior is not RHF induced, but rather the way Zod works. Check out Zod issue thread: colinhacks/zod#479
> the `.refine()` function defined on the object type doesn't get triggered until the all fields in the object are defined

**Reply by apn-atobe** [reply]
So this is a spec... I learned a lot.
Thanks for the reply!

**Comment by leapful** [comment, marked as answer]
@pcp-atobe : Your zod schema defines "NumberInput" as a mandatory field therefore when parsing and validating data by zod library, the target data is not valid and stops at validation phase instead of continuing to "refinement" phase.

Switching "NumberInput" to optional will solve the issue. Working demo: https://codesandbox.io/s/nostalgic-sara-i01e94?file=/src/App.js:364-372

**Reply by apn-atobe** [reply]
Thank you for editing the sandbox to give a clear example.
It was helpful!

**Comment by jmons54** [comment]
Use partial : `object().partial().refine()`

https://codesandbox.io/s/blissful-wright-33wc6w?file=/src/App.js

**Comment by kibaek-kimm** [comment]
If use `partial` method, all required fields are change to optional.

I fixed to a new group for password.

```typescript
const reqBodyScheme = z
  .object({
    passwordGroup: z.object({
      password: z.string().min(1, { message: "required" }),
      passwordConfirm: z.string().min(1, { message: "required" })),
    })
    .refine(
      (data) => {
        console.log(data);
        return data.password === data.passwordConfirm;
      },
      {
        message: "Passwords don't match",
        path: ["passwordConfirm"], }
      }
    ),
    NumberInput: z.number() // <- If you comment this out and use only z.string(), the refine works
  });
;
```

**Reply by zuracode** [reply]
best answer, I tried many things which was mainly offered from another github issues but none of them worked. Thank you so much
